Add genre director profile views and react router

mdillemuth opened the pull request:
I have a few problems so far with this branch and I believe they are mainly about the 'user' state. 

![add-director-view-genre-view](https://user-images.githubusercontent.com/67530427/100957826-310ea600-34e9-11eb-9206-3785a8210862.gif)

**Things that work**
- Registering a new account
- Removing an account
- Logging in
- Logging out
- Displaying favorites in profile view
- Removing a favorite from profile view
- Displaying movies with same director in director view
- Displaying movies with same genre in genre view

**Things that DONT work**
- The moviecards sometimes do not load immediately after logging in, this is shown in the gif
- Updating account information from profile view
- Refreshing the page while in profile view 
- Immediately updating the user's favorites after removing a favorite while in profile view

mdillemuth committed:
Add login routing and store user token in local storage

mdillemuth committed:
Add logout feature

mdillemuth committed:
Add client-side routing throughout app

mdillemuth committed:
Add routing to directorview, genreview, and profileview

mdillemuth committed:
Add 'similar movies' section in directorview and genreview

mdillemuth committed:
Add user favorite moviecards to profileview

mdillemuth committed:
Add user registration backend functionality

mdillemuth committed:
Add delete account function to profileView

mdillemuth committed:
Add remove favorite movie function to profileView

mdillemuth committed:
Add styling for director & genre views, update moviecard styling

mdillemuth committed:
Add link route to NavBar component

eadsimone commented on src/components/profile-view/profile-view.jsx:
```diff
@@ -0,0 +1,214 @@
+    password: '',
+    birthday: '',
+  });
+
```
`
componentDidMount() {
		const accessToken = localStorage.getItem('token');
		if (accessToken !== null) {
			this.getUser(accessToken);
		}
	}`

	getUser(token) {
		const username = localStorage.getItem('user');
		axios
			.get(`https://bchanmyflix.herokuapp.com/users/${username}`, {
				headers: { Authorization: `Bearer ${token}` },
			})
			.then((response) => {
				this.setState({
					Username: response.data.Username,
					Password: response.data.Password,
					Email: response.data.Email,
					Birthday: response.data.Birthday,
					FavoriteMovies: response.data.FavoriteMovies,
				});
			})
			.catch(function (error) {
				console.log(error);
			});
	}

eadsimone commented on src/components/main-view/main-view.jsx:
```diff
@@ -1,25 +1,43 @@
+
+    localStorage.setItem('token', authData.token);
+    localStorage.setItem('user', authData.user.Username);
+  };
```
this.getMovies(authData.token);

mdillemuth committed:
Add getUsers() to fetch user data from API

mdillemuth committed:
Recreate profileview with user data from API, add form validation

mdillemuth committed:
profileView receives movies from API call, no longer passed down from mainview as prop

mdillemuth committed:
Refactor getFavoriteMovies to work in profileView

mdillemuth committed:
Refactor profileView code and implement conditional rendering for favorite movies

mdillemuth committed:
Refactor favoriteMovies handling

mdillemuth committed:
Refactor updateAccount put request headers

mdillemuth committed:
Make updateAccount work from profileView, remove faulty form validation

eadsimone commented on src/components/movie-view/movie-view.jsx:
```diff
@@ -9,45 +11,55 @@ class MovieView extends Component {
+                  {movie.Title || ''}
+                </span>
+                <p className='h4 text-dark ml-2'>
+                  <i className='star fa fa-star-o text-warning'></i>
```
for favorites you can something like:
```
<div className="favorites-button-container">
								<Button className="movie-view-button favorites-button" value={movie._id} onClick={(e) => this.handleAddFavorite(e, movie)}>
									Add to Favorites
								</Button>
							</div>
```

eadsimone commented on src/components/movie-view/movie-view.jsx:
```diff
@@ -9,45 +11,55 @@ class MovieView extends Component {
   }
 
```
```
handleAddFavorite(e, movie) {
		e.preventDefault();
		const username = localStorage.getItem('user');
		const token = localStorage.getItem('token');
		axios({
			method: 'post',
			url: `https://bchanmyflix.herokuapp.com/users/${username}/Movies/${movie._id}`,
			headers: { Authorization: `Bearer ${token}` },
		})
			.then(() => {
				alert(`${movie.Title} was add to Favorites`);
			})
			.catch(function (err) {
				console.log(err);
			});
	}
```

mdillemuth committed:
Implement add & remove favorite movie

mdillemuth committed:
Add toggleFavorite feature, add alert confirmation to removeAccount

mdillemuth committed:
Add client-side form validation for login, registration, and update

mdillemuth committed:
Confirm passwords match on form submit  for registration and update

mdillemuth committed:
Add server-side validation UI alert for login view

mdillemuth committed:
Add server-side validation UI alert to registration view, remove unused style sheets in components/

mdillemuth committed:
Add and update propTypes

mdillemuth committed:
Fix alert display for multiple failed login attempts, add loading spinner on login

mdillemuth committed:
Add loading spinner to registration, fix undefined behavior in moviecard

mdillemuth committed:
Add seperate form component for login/registration

mdillemuth committed:
Create separate form component for profileView, add propTypes to Navbar

mdillemuth committed:
Refactor form props to clean up code

mdillemuth committed:
Create separate components for displaying favorite movies and all movies

mdillemuth committed:
Refactor mainView routes, add blank navbar to loginView, change loading gif

mdillemuth committed:
Add favicon and update title

mdillemuth committed:
Add not-found component and redirect routing

mdillemuth commented:
![3-5-gif-final](https://user-images.githubusercontent.com/67530427/101987303-47430180-3c61-11eb-9fab-885a1bf022e5.gif)

**Major Changes**
- Client-side authentication that prevents submitting invalid forms to the server
- Server-side authentication that alerts the user when the received form was invalid (ex. username taken)
- Redirecting from invalid URLs to a '/not-found' page
- Code refactoring and abstracting components to clean-up project
  - Separate components for registration form and login form
  - Separate component for displaying favorites in profileView (favoritesView)
  - Separate component for displaying movies 
-----
**Minor Changes**
  - Added favicon
  - Added customized loading spinner gif for when form is submitted to display while waiting for server response
  - NavBar is abstracted to render on all routes, but profile/logout buttons only render when user after user is logged in

eadsimone reviewed (approved):
![urock](https://user-images.githubusercontent.com/2244503/101988388-04d8ef00-3c78-11eb-88cf-09f01290ce93.gif)
